nni/common/concrete_trace_utils/concrete_tracer.py

```diff
@@ -0,0 +1,1256 @@
+                raise Exception('more than 1 tracer detected. please report the issue')
+        def __eq__(self, __o: object) -> bool:
+            return id(__o) in (id(self), id(clz))
+        def __hash__(self):
+            return id(self)
+    return clz_wrapper_clz()
+
+def _create_wrapped_leaf_iterable_class(tracer: ConcreteTracer, clz):
+    class clz_wrapper_clz:
+        @functools.wraps(clz)
+        def __call__(self, *args, **kwargs):
+            if tracer.temp_disable_call:
+                return clz(*args, **kwargs)
+            tracers = _orig_set()
+            if _orig_len(args) != 0:
+                if _orig_isinstance(args[0], ep.Proxy):
+                    tracers.add(args[0].tracer)
+                if _orig_isinstance(args[0], Iterator):
+                    args = (clz(args[0]), *args[1:])
+                if _orig_isinstance(args[0], Iterable):
+                    for item in args[0]:
+                        if _orig_isinstance(item, ep.Proxy):
+                            tracers.add(item.tracer)
+            if _orig_len(tracers) == 0:
+                return clz(*args, **kwargs)
+            elif _orig_len(tracers) == 1 and next(iter(tracers)) == tracer:
+                return tracer.create_proxy('call_function',
+                    clz, args, kwargs)
+            else:
+                raise Exception('more than 1 tracer detected. please report the issue')
+        def __eq__(self, __o: object) -> bool:
+            return id(__o) in (id(self), id(clz))
+        def __hash__(self):
+            return id(self)
+    clz_wrapper = clz_wrapper_clz()
+    for name in dir(clz):
+        attr = _orig_getattr(clz, name)
+        if not name.startswith('_') or name in ('__getitem__', '__setitem__', '__iter__', '__len__'):
+            if _orig_isinstance(attr, Callable):
+                setattr(clz_wrapper, name, _create_wrapped_leaf_method(tracer, attr, name, None))
+            else:
+                setattr(clz_wrapper, name, attr)
+    return clz_wrapper
+
+def _create_wrapped_attr_for_middle_class(tracer: ConcreteTracer, clz, the_path_of_middle_class):
+    _orig_clz_getattribute = clz.__getattribute__
+    if hasattr(clz, '__getattr__'):
+        _orig_clz_getattr = clz.__getattr__
+    else:
+        _orig_clz_getattr = None
+    @functools.wraps(_orig_clz_getattribute)
+    def clz_getattr_wrapper(obj, attr):
+        if tracer.temp_disable_call | tracer.temp_disable_attr:
+            if _orig_clz_getattr == None:
+                return _orig_clz_getattribute(obj, attr)
+            else:
+                try:
+                    return _orig_clz_getattribute(obj, attr)
+                except AttributeError:
+                    return _orig_clz_getattr(obj, attr)
+        else:
+            return tracer.create_proxy('get_attr', f'{the_path_of_middle_class[id(obj)]}.{attr}', (), {})
+    return clz_getattr_wrapper
+
+def concrete_trace(root : Union[torch.nn.Module, Callable[..., Any]],
```

Comment:
This is the top-level API, right?

I think it deserves some docstrings to explain:

- Who should use this?
- What are its expected input arguments?
- What is its return type? How to interpret / use its return value?